### PR TITLE
LinalgExt: remove EncodingAttr, redundant with TensorEncodingAttr

### PR DESCRIPTION
**Pull request by bjacob**

### CODE REVIEW

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtBase.td [hunk 1]
```diff
@@ -97,42 +97,10 @@ def TensorEncodingEnum
...
 
 def TensorEncodingAttr :
-  EnumAttr<IREELinalgExt_Dialect, TensorEncodingEnum, ""> {
-  let assemblyFormat = "``$value";
-}
-
-def IREELinalgExt_EncodingAttr : IREELinalgExt_Attr<"Encoding"> {
-  let mnemonic = "encoding";
-  let summary = [{tensor layout encoding}];
-  let description = [{
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgExt/IR/LinalgExtBase.td:107]: I think we want to keep the description.

**Review by hanhanW — APPROVED** [review]

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp [hunk 2]
```diff
@@ -255,7 +255,11 @@ summarizeDispatchWorkgroupsOp(DispatchWorkgroupsOp regionOp) {
       })
       .Case<IREE::LinalgExt::UnsetEncodingOp>([&](auto op) {
         auto opName = getOpNameWithoutDialectName(op);
-        auto encoding = stringifyEnum(op.getSourceTensorEncoding());
+        auto encoding = stringifyEnum(
+            op.getSourceType()
+                .getEncoding()
+                .template cast<IREE::LinalgExt::TensorEncodingAttr>()
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp:261]: This is not a template method, do we need the `template` keyword here?

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/MaterializeEncoding.cpp [hunk 3]
```diff
@@ -27,11 +27,16 @@ using namespace mlir::iree_compiler::IREE::LinalgExt;
 //===---------------------------------------------------------------------===//
 
 /// Extract encoding from the `tensorType` if specified.
-static std::optional<TensorEncoding> getEncoding(RankedTensorType tensorType) {
-  auto encodingAttr = tensorType.getEncoding().dyn_cast_or_null<EncodingAttr>();
-  if (!encodingAttr)
+std::optional<TensorEncoding> getEncoding(RankedTensorType tensorType) {
```
Comment (hanhanW) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/MaterializeEncoding.cpp]: shouldn't this be static method?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp (hunk 2 above)
Comment (bjacob) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp:261]: Clang complained about that. I think it's correct. Notice the `auto op` parameter to the lambda (at line 256). Defining a lambda taking an `auto` parameter makes it implicitly templated. This is a C++14 feature. See https://en.cppreference.com/w/cpp/language/lambda,

> For each parameter in params whose type is specified as auto, an invented template parameter is added to template-params, in order of appearance.

Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp:261]: I see, I did not check the lambda things, sorry about that and thanks for the reply!

**Pull request merged**